crates/holochain_persistence_lmdb/src/txn/lmdb.rs

```diff
@@ -0,0 +1,231 @@
+                    let map_size = env_lock.info().map_err(to_api_error)?.map_size();
+                    env_lock.set_map_size(map_size * 2).map_err(to_api_error)?;
+                    Ok(false)
+                }
+                r => Err(to_api_error(r)), // preserve any other errors
+            }
+        })
+    }
+}
+
+impl<A: Attribute + Sync + Send + DeserializeOwned> holochain_persistence_api::txn::Writer
+    for EnvCursor<A>
+{
+    fn commit(self) -> PersistenceResult<()> {
+        loop {
+            let ret = self.commit_internal()?;
+            if ret {
+                return Ok(());
+            }
+        }
+    }
+}
+
+impl<A: Attribute> ReportStorage for EnvCursor<A> {
+    fn get_storage_report(&self) -> PersistenceResult<StorageReport> {
+        Ok(StorageReport::new(0)) // TODO: implement this
+    }
+}
+
+impl<A: Attribute> EnvCursor<A> {
+    pub fn new(
+        cas_db: LmdbStorage,
+        eav_db: EavLmdbStorage<A>,
+        staging_cas_db: LmdbStorage,
+        staging_eav_db: EavLmdbStorage<A>,
+    ) -> Self {
+        Self {
+            cas_db,
+            eav_db,
+            staging_cas_db,
+            staging_eav_db,
+            phantom: std::marker::PhantomData,
+        }
+    }
+}
+
+fn to_api_error(e: rkv::error::StoreError) -> PersistenceError {
+    // Convert to lmdb persistence error
+    let e: crate::error::PersistenceError = e.into();
+
+    // Convert into api persistence error
+    let e: PersistenceError = e.into();
+    e
+}
+
+impl<A: Attribute> ContentAddressableStorage for EnvCursor<A> {
+    fn add(&self, content: &dyn AddressableContent) -> PersistenceResult<()> {
+        self.staging_cas_db.add(content)
+    }
+
+    fn contains(&self, address: &Address) -> PersistenceResult<bool> {
+        self.fetch(address)
+            .map(|maybe_content| maybe_content.is_some())
+    }
+
+    fn fetch(&self, address: &Address) -> PersistenceResult<Option<Content>> {
+        let maybe_content = self.staging_cas_db.fetch(address)?;
+
+        if maybe_content.is_some() {
+            return Ok(maybe_content);
+        }
+
+        let maybe_content = self.cas_db.fetch(address)?;
+
+        if let Some(content) = maybe_content {
+            self.staging_cas_db.add(&content)?;
+            Ok(Some(content))
+        } else {
+            Ok(None)
+        }
+    }
+
+    fn get_id(&self) -> uuid::Uuid {
+        self.cas_db.get_id()
+    }
+}
+
+impl<A: Attribute + serde::de::DeserializeOwned> EntityAttributeValueStorage<A> for EnvCursor<A> {
+    fn add_eavi(
+        &self,
+        eav: &EntityAttributeValueIndex<A>,
+    ) -> PersistenceResult<Option<EntityAttributeValueIndex<A>>> {
+        self.staging_eav_db.add_lmdb_eavi(eav).map_err(to_api_error)
+    }
+
+    fn fetch_eavi(
+        &self,
+        query: &EaviQuery<A>,
+    ) -> PersistenceResult<BTreeSet<EntityAttributeValueIndex<A>>> {
+        let eavis = self.staging_eav_db.fetch_eavi(query)?;
+
+        if !eavis.is_empty() {
```

Comment:
I stumbled over this line of code...

```suggestion
    use std::ops::Not;
        if eavis.is_empty().not() {
```